File: projects/chapter12_inheritance/main/array_allocation_test/main.cpp
#include <iostream>
#include <chrono>
using namespace std;
using namespace chrono;

const int ARRAY_SIZE = 10000;
const int ITERATIONS = 100000;

// Static array (allocated at compile time, stored in data segment)
int static_array[ARRAY_SIZE];

void test_static_array() {
    for(int i = 0; i < ARRAY_SIZE; i++) {
        static_array[i] = i;
    }
}

// Stack array (allocated on function call, stored in stack)
void test_stack_array() {
    int stack_array[ARRAY_SIZE];
    for(int i = 0; i < ARRAY_SIZE; i++) {
        stack_array[i] = i;
    }
}

// Heap array (allocated dynamically at runtime)
void test_heap_array() {
    int* heap_array = new int[ARRAY_SIZE];
    for(int i = 0; i < ARRAY_SIZE; i++) {
        heap_array[i] = i;
    }
    delete[] heap_array;
}

int main() {
    cout << "🎉 Let's race some arrays! 🏁" << endl;
    cout << "Array size: " << ARRAY_SIZE << " (that's a chonky boi!)" << endl;
    cout << "Iterations: " << ITERATIONS << " (buckle up!)" << endl << endl;

    // Test static array
    auto start = high_resolution_clock::now();
    for(int i = 0; i < ITERATIONS; i++) {
        test_static_array();
    }
    auto end = high_resolution_clock::now();
    auto static_duration = duration_cast<milliseconds>(end - start);
    cout << "⚡ Static array time: " << static_duration.count() << " ms (ZOOOOOM!)" << endl;

    // Test stack array
    start = high_resolution_clock::now();
    for(int i = 0; i < ITERATIONS; i++) {
        test_stack_array();
    }
    end = high_resolution_clock::now();
    auto stack_duration = duration_cast<milliseconds>(end - start);
    cout << "📚 Stack array time: " << stack_duration.count() << " ms (pretty speedy!)" << endl;

    // Test heap array
    start = high_resolution_clock::now();
    for(int i = 0; i < ITERATIONS; i++) {
        test_heap_array();
    }
    end = high_resolution_clock::now();
    auto heap_duration = duration_cast<milliseconds>(end - start);
    cout << "🗑️  Heap array time: " << heap_duration.count() << " ms (taking its sweet time...)" << endl;

    cout << endl << "📊 The Verdict:" << endl;
    cout << "⚡ Static: Allocated once at compile time (data segment)" << endl;
    cout << "   BUT: Can be slower due to cache misses and memory contention!" << endl;
    cout << "📚 Stack: Often FASTEST! Stack pointer moves = instant allocation" << endl;
    cout << "   PLUS: Hot in CPU cache from recent use = speedy access!" << endl;
    cout << "🗑️  Heap: new/delete overhead, but modern allocators are optimized" << endl;
    cout << "   Cache locality on fresh allocations can beat static!" << endl;
    cout << endl << "💡 Plot twist: Stack often wins when you're writing to arrays!" << endl;
    cout << "   Theory said static would dominate, but cache locality matters more!" << endl;

    return 0;
}

File: projects/chapter12_inheritance/main/array_allocation_test/Main.java
public class Main {
    static final int ARRAY_SIZE = 10000;
    static final int ITERATIONS = 100000;

    // Static array (class-level, allocated once when class loads)
    static int[] static_array = new int[ARRAY_SIZE];

    static void testStaticArray() {
        for(int i = 0; i < ARRAY_SIZE; i++) {
            static_array[i] = i;
        }
    }

    // In Java, all arrays are heap-allocated
    // This simulates "local array" behavior
    static void testLocalArray() {
        int[] local_array = new int[ARRAY_SIZE];
        for(int i = 0; i < ARRAY_SIZE; i++) {
            local_array[i] = i;
        }
    }

    // Heap array with explicit creation each time
    static void testHeapArray() {
        int[] heap_array = new int[ARRAY_SIZE];
        for(int i = 0; i < ARRAY_SIZE; i++) {
            heap_array[i] = i;
        }
        // Java handles deallocation via garbage collection
    }

    public static void main(String[] args) {
        System.out.println("🎉 Let's race some arrays! 🏁");
        System.out.println("Array size: " + ARRAY_SIZE + " (that's a chonky boi!)");
        System.out.println("Iterations: " + ITERATIONS + " (buckle up!)");
        System.out.println();

        // Test static array
        long start = System.nanoTime();
        for(int i = 0; i < ITERATIONS; i++) {
            testStaticArray();
        }
        long end = System.nanoTime();
        long static_duration = (end - start) / 1_000_000;
        System.out.println("⚡ Static array time: " + static_duration + " ms (ZOOOOOM!)");

        // Test local array
        start = System.nanoTime();
        for(int i = 0; i < ITERATIONS; i++) {
            testLocalArray();
        }
        end = System.nanoTime();
        long local_duration = (end - start) / 1_000_000;
        System.out.println("☕ Local array time: " + local_duration + " ms (Java doing Java things)");

        // Test heap array
        start = System.nanoTime();
        for(int i = 0; i < ITERATIONS; i++) {
            testHeapArray();
        }
        end = System.nanoTime();
        long heap_duration = (end - start) / 1_000_000;
        System.out.println("🗑️  Heap array time: " + heap_duration + " ms (same same but different)");

        System.out.println();
        System.out.println("📊 The Verdict:");
        System.out.println("⚡ Static: WINNER! Allocated once when class loads (efficiency king!)");
        System.out.println("☕ Local: Pretty much heap - Java loves its heap");
        System.out.println("🗑️  Heap: Also heap - surprise! Everything's heap in Java!");
        System.out.println("💡 Fun fact: Java doesn't do stack arrays. It's all heap, baby!");
    }
}
